Aarch64 test errors

Hey as I'm packaging the nutils package for nixos I found that the tests on aarch64 are failing. 
https://hydra.nixos.org/build/176018982/nixlog/2
```
tests/test_evaluable.py:59: in assertFunctionAlmostEqual
    self.assertArrayAlmostEqual(actual.eval(**evalargs), desired, decimal)
tests/test_evaluable.py:54: in assertArrayAlmostEqual
    self.fail(''.join(lines))
E   AssertionError: arrays are not equal up to 10 decimals in 9/9 entries:
E     (0, 0) actual=-1.8014398509481932e+16 desired=1.8014398509482092e+16 difference=-3.6028797018964024e+16
E     (0, 1) actual=3.6028797018963868e+16 desired=-3.602879701896421e+16 difference=7.205759403792808e+16
E     (0, 2) actual=-1.8014398509481932e+16 desired=1.8014398509482092e+16 difference=-3.6028797018964024e+16
E     (1, 0) actual=3.602879701896387e+16 desired=-3.602879701896419e+16 difference=7.205759403792806e+16
E     (1, 1) actual=-7.205759403792776e+16 desired=7.205759403792843e+16 difference=-1.441151880758562e+17
E     (1, 2) actual=3.602879701896387e+16 desired=-3.602879701896419e+16 difference=7.205759403792806e+16
E     (2, 0) actual=-1.801439850948193e+16 desired=1.8014398509482092e+16 difference=-3.6028797018964024e+16
E     (2, 1) actual=3.602879701896387e+16 desired=-3.602879701896421e+16 difference=7.205759403792808e+16
E     (2, 2) actual=-1.801439850948193e+16 desired=1.8014398509482092e+16 difference=-3.6028797018964024e+16
=============================== warnings summary ===============================
``

--- a/nutils/evaluable.py
+++ b/nutils/evaluable.py
@@ -2356,6 +2356,34 @@ class ArcTan(Pointwise):
     complex_deriv = lambda x: reciprocal(1+x**2),
 
 
+class CosH(Pointwise):
+    'Hyperbolic cosine, element-wise.'
+    __slots__ = ()
+    evalf = numpy.cosh
+    complex_deriv = lambda x: SinH(x),
+
+
+class SinH(Pointwise):
+    'Hyperbolic sine, element-wise.'
+    __slots__ = ()
+    evalf = numpy.sinh
+    complex_deriv = CosH,
+
+
+class TanH(Pointwise):
+    'Hyperbolic tangent, element-wise.'
+    __slots__ = ()
+    evalf = numpy.tanh
+    complex_deriv = lambda x: 1 - TanH(x)**2,
+
+
+class ArcTanH(Pointwise):
+    'Inverse tangent, element-wise.'
+    __slots__ = ()
+    evalf = numpy.arctanh
+    complex_deriv = lambda x: reciprocal(1-x**2),
+
+
 class Exp(Pointwise):
     __slots__ = ()
     evalf = numpy.exp
@@ -4365,19 +4393,19 @@ def abs(arg):
 
 
 def sinh(arg):
-    return .5 * (exp(arg) - exp(-arg))
+    return SinH(arg)
 
 
 def cosh(arg):
-    return .5 * (exp(arg) + exp(-arg))
+    return CosH(arg)
 
 
 def tanh(arg):
-    return 1 - 2. / (exp(2*arg) + 1)
+    return TanH(arg)
 
 
 def arctanh(arg):
-    return .5 * (ln(1+arg) - ln(1-arg))
+    return ArcTanH(arg)
 
 
 def divide(arg1, arg2):

--- a/nutils/function.py
+++ b/nutils/function.py
@@ -1220,7 +1220,7 @@ def subtract(__left: IntoArray, __right: IntoArray) -> Array:
     :class:`Array`
     '''
 
-    return add(__left, negative(__right))
+    return _Wrapper.broadcasted_arrays(evaluable.subtract, __left, __right)
 
 
 @implements(numpy.positive)
@@ -1252,7 +1252,7 @@ def negative(__arg: IntoArray) -> Array:
     :class:`Array`
     '''
 
-    return multiply(__arg, -1)
+    return _Wrapper.broadcasted_arrays(evaluable.negative, __arg)
 
 
 @implements(numpy.multiply)
@@ -1322,7 +1322,7 @@ def reciprocal(__arg: IntoArray) -> Array:
     :class:`Array`
     '''
 
-    return power(__arg, -1.)
+    return _Wrapper.broadcasted_arrays(evaluable.reciprocal, __arg)
 
 
 @implements(numpy.power)
@@ -1354,7 +1354,7 @@ def sqrt(__arg: IntoArray) -> Array:
     :class:`Array`
     '''
 
-    return power(__arg, .5)
+    return _Wrapper.broadcasted_arrays(evaluable.sqrt, __arg, min_dtype=float)
 
 
 @implements(numpy.square)
@@ -1596,8 +1596,7 @@ def cosh(__arg: IntoArray) -> Array:
     :class:`Array`
     '''
 
-    arg = Array.cast(__arg)
-    return .5 * (exp(arg) + exp(-arg))
+    return _Wrapper.broadcasted_arrays(evaluable.cosh, __arg, min_dtype=float)
 
 
 @implements(numpy.sinh)
@@ -1613,8 +1612,7 @@ def sinh(__arg: IntoArray) -> Array:
     :class:`Array`
     '''
 
-    arg = Array.cast(__arg)
-    return .5 * (exp(arg) - exp(-arg))
+    return _Wrapper.broadcasted_arrays(evaluable.sinh, __arg, min_dtype=float)
 
 
 @implements(numpy.tanh)
@@ -1630,8 +1628,7 @@ def tanh(__arg: IntoArray) -> Array:
     :class:`Array`
     '''
 
-    arg = Array.cast(__arg)
-    return 1 - 2. / (exp(2*arg) + 1)
+    return _Wrapper.broadcasted_arrays(evaluable.tanh, __arg, min_dtype=float)
 
 
 @implements(numpy.arctanh)
@@ -1647,8 +1644,7 @@ def arctanh(__arg: IntoArray) -> Array:
     :class:`Array`
     '''
 
-    arg = Array.cast(__arg)
-    return .5 * (ln(1+arg) - ln(1-arg))
+    return _Wrapper.broadcasted_arrays(evaluable.arctanh, __arg, min_dtype=float)
 
 
 @implements(numpy.exp)
@@ -1699,7 +1695,7 @@ def log2(__arg: IntoArray) -> Array:
     :class:`Array`
     '''
 
-    return log(__arg) / log(2)
+    return _Wrapper.broadcasted_arrays(evaluable.log2, __arg, min_dtype=float)
 
 
 @implements(numpy.log10)
@@ -1715,7 +1711,7 @@ def log10(__arg: IntoArray) -> Array:
     :class:`Array`
     '''
 
-    return log(__arg) / log(10)
+    return _Wrapper.broadcasted_arrays(evaluable.log10, __arg, min_dtype=float)
 
 # COMPARISON
 

--- a/tests/test_evaluable.py
+++ b/tests/test_evaluable.py
@@ -408,6 +408,8 @@ def generate(*shape, real, imag, zero, negative):
         raise Exception('shape is too small to test at least one of all selected number categories')
     if real or imag:
         a = numpy.tanh(2 * a / a[-1])  # map to (-1,1)
+        if negative:
+            a[:iz] -= a[iz-1] / 2 # introduce asymmetry to reduce risk of singular matrices
         if real and imag:
             assert negative
             a = a * numpy.exp(1j * numpy.arange(size)**2)

